Results-discriminator model and Model Calibration (#15)
greenelab/text_mined_hetnet_manuscript: content/04.results.md

## Random Sampling of Generative Model
place the grid aurocs here for generative model

### Discriminator Model Builds Off Generative Model

> We evaluate the discriminator model's ability to build off of the generative model.

Reviewer: ```suggestion
In the framework we are using, a generative model is trained over label functions to produce probabilistic training labels for each sentence.
Then we train a discriminative model, which has full access to a representation of the text of the sentence, to predict the generated labels.
```

> Each generative model, trained on randomly sampled label functions, generates probabilistic training labels for each sentence in our dataset.

Reviewer: ```suggestion
```

> For example, one sentence could receive a high confidence score (>=0.6), a low score (0.49<=) or it could receive a borderline score of 0.5.

Reviewer: ```suggestion
```

> Using these confidence scores, the discriminator model constructs sentence features and generates a final confidence score (0-1) for each sentence in our dataset.

Reviewer: ```suggestion
```

> We report the results of the discriminative model using area under the precision recall curve (AUPRC) (Figure {@fig:discriminative_model_auprc}) and area under the receiver operating curve (AUROC) (Figure {@fig:discriminative_model_auroc}).    

Reviewer: ```suggestion
The discriminative model is a convolutional neural network trained over word embeddings.
We report the results of the discriminative model using area under the precision recall curve (AUPRC) (Figure {@fig:discriminative_model_auprc}) and area under the receiver operating curve (AUROC) (Figure {@fig:discriminative_model_auroc}).    
```

> Overall the discriminative model performs well when using a generative model that is trained off of edge specific label functions. 

Reviewer: Wait - isn't the disciminative model usually worse than the generative model? Am I misreading the results.

> In the top left corner of figures {@fig:discriminative_model_auroc} and {@fig:discriminative_model_auprc} the discriminative model underperforms the generative model when only using distant supervision label functions.

Reviewer: ```suggestion
In the top left corner of figures {@fig:discriminative_model_auroc} and {@fig:discriminative_model_auprc} the discriminative model underperforms the generative model when only using distant supervision label functions, but the discriminative model surpasses the generative model when all label functions are included.
```

> The discriminative model over takes the generative model as more label functions are incorporated. 

Reviewer: ```suggestion
```

> This similar trend is seen for the Compound treats Disease (CtD) edge type; however, both the generative and discriminative model perform equally instead of one outcompeting another.

Reviewer: I'm going to hold off continuing to review this section in case I'm misreading the results. Can you take a look and see if your interpretation matches mine (the discriminative model usually doesn't help, and is often worse)?

Interestingly, the discriminative model does well in predicting the CbG edge type, when using the generative model trained on GiG label functions.
This trend doesn't hold for the reverse as the discriminative model slightly underperforms the generative model.
Lastly for CbG and GiG using all label functions didn't aid in the discriminative model performance.

![
Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models.
The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. 
For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. 
The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship.
This pattern continues filling out the rest of the grid.

> The last most column consists of pooling every relationship specific label function and proceeding as above.

Reviewer: Put in what the error bars represent. You should do this in all figures with error bars.

![
Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models.
The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. 
For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. 
The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship.
This pattern continues filling out the rest of the grid.
The last most column consists of pooling every relationship specific label function and proceeding as above.
](https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/label_sampling_experiment/disc_performance_test_set_auprc.png){#fig:discriminative_model_auprc}

### Discriminative Model Calibration

> Deep learning models can be overconfident in their predictions. 

Reviewer: ```suggestion
Even deep learning models with high precision and recall can be poorly calibrated, and the overconfidence of these models has been noted [@arxiv:1706.04599; @arxiv:1807.00263].
```

> Based on this notion we calibrated the best performing discriminative model using a temperature scaling algorithm.

Reviewer: ```suggestion
We attempted to calibrate the best performing discriminative model so that we could directly use the emitted probabilities.
```

> For Disease associates Gene (DaG) and Compound treats Disease (CtD) the model is calibrated for high prob predictions.

Reviewer: ```suggestion
We examined the calibration of our existing model (Figure {@fig:discriminative_model_calibration}, blue line).
We found that the DaG and CtG edge types were, though not perfectly calibrated, were somewhat aligned with the ideal calibration lines.
The CbG and GiG edges were poorly calibrated and increasing model certainty did not always lead to an increase in precision.
Applying the calibration algorithm (orange line) did not appear to bring predictions in line with the ideal calibration line, but did capture some of the uncertainty in the GiG edge type.
For this reason we use the measured precision instead of the predicted probabilities when determining how many edges could be added to existing knowledge bases with specified levels of confidence.
```

> For Compound binds Gene (CbG) and Gene interacts Gene (GiG) the model does not calibrate well.

Reviewer: ```suggestion
```

> The CbG model has the predicted probabilities shrinking into a domain of [0.2-0.6].

Reviewer: ```suggestion
```

> The GiG model predictions were overestimated so the calibrated range shrink to be between 0.4 and 0.6.

Reviewer: ```suggestion
```

## Random Sampling of Generative Model
place the grid aurocs here for generative model

> ### Discriminator Model Builds Off Generative Model

Reviewer: ```suggestion
### Discriminative Model Performance

```

In this framework we used a generative model trained over label functions to produce probabilistic training labels for each sentence.
Then we train a discriminative model, which has full access to a representation of the text of the sentence, to predict the generated labels.
The discriminative model is a convolutional neural network trained over word embeddings.

> We report the results of the discriminative model using area under the precision recall curve (AUPR) (Figure {@fig:discriminative_model_auprc}) and area under the receiver operating curve (AUROC) (Figure {@fig:discriminative_model_auroc}).    

Reviewer: Did you already introduce these earlier in the manuscript? If so, no need to re-define them.

> Generally, the discriminative model under-performs compared to the generative model.

Reviewer: ```suggestion
We found that the discriminative model under-performed the generative model in most cases.
```

> In the top left corner of figures {@fig:discriminative_model_auroc} and {@fig:discriminative_model_auprc} the discriminative model underperforms the generative model when only using distant supervision label functions.

Reviewer: ```suggestion
Only for the CtD edge does the discriminative model appear to provide performance above that of the generative model, and that increased performance is only with modest numbers of label functions.
With the full set of label functions, the performance of both remains similar.
The trend observed in the generative model that one or a few mismatched label functions (off-diagonal) improves performance is retained despite the limited performance of the discriminative model.
```

> As the number of label functions increase the discriminative model starts to have similar performance to the generative model.

Reviewer: ```suggestion
```

> Compound-treats-Disease (CtD) starts off with the opposite trend.

Reviewer: ```suggestion
```

> The discriminative model outcompetes the generative model; however, as more label functions are added performance starts to decline.

Reviewer: ```suggestion
```

> For the Compound-binds-Gene (CbG) edge type, the discriminative model consistently underperforms the generative model.

Reviewer: ```suggestion
```

> Gene-interacts-Gene (GiG) has a similar pattern to DaG.

Reviewer: ```suggestion
```

> The discriminative model starts off underperforming the generative model.

Reviewer: ```suggestion
```

> As soon as more label functions are incorporated performance improves to be on par with the generative model.  

Reviewer: ```suggestion
```

Deleting a ton of this. You don't need to describe every panel of the figure (that'd take a thousand words, since an image is worth a thousand words 😉 ). You need to give the main take-home points and cite the figure for support.

> With respect to the off-diagonals, the discriminative model uses annotations from the generative model, that is trained using label functions other than the targeted edge type (targeted edge type shown on the y-axes of figures {@fig:discriminative_model_auroc} and {@fig:discriminative_model_auprc}). 

Reviewer: ```suggestion
```

> Overall, performance for the discriminative model is second-rate. 

Reviewer: ```suggestion
```

> For the top row the discriminative model under performs the generative model.

Reviewer: ```suggestion
```

> When using CtD label functions the auroc dips as low as 50%.

Reviewer: ```suggestion
```

Reviewer: If this was being kept, AUROC should have been capitalized as it's an acronym. Worth checking the rest of the paper to see if you have it in caps.

> Similar scores can be seen with CbG and All (Figure {@fig:discriminative_model_auprc}). 

Reviewer: ```suggestion
```

> For the second row, there is an inverse effect.

Reviewer: ```suggestion
```

> The discriminative model starts off out-performing the generative model, but diminishes as the number of label functions increases.

Reviewer: ```suggestion
```

> In the last two rows performance spikes when using one label function, but plateaus or sharply declines as more are added.

Reviewer: ```suggestion
```

> Despite the decreasing trend, the discriminative model performs almost on par with the generative model when predicting the CbG edge type using GiG label functions.

Reviewer: ```suggestion
```

> The same trend can be found when predicting the GiG edge using CbG label functions.

Reviewer: ```suggestion
```

> Lastly only CtD and DaG slightly benefited from using all label functions at once.

Reviewer: ```suggestion
```

> For CbG and GiG they have performance increase using one label function then plateaus or declines when more are added.

Reviewer: ```suggestion
```

![
Grid of Area Under the Receiver Operating Curve (AUROC) scores for each discriminative model trained using generated labels from the generative models.
The rows depict the relationship each model is trying to predict and the columns are the relationship specific sources each label function was sampled from. 
For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. 
The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship.
This pattern continues filling out the rest of the grid.
The last most column consists of pooling every relationship specific label function and proceeding as above.
](https://raw.githubusercontent.com/greenelab/snorkeling/master/figures/label_sampling_experiment/disc_performance_test_set_auprc.png){#fig:discriminative_model_auprc}

> ### Discriminative Model Calibration

Reviewer: ```suggestion
### Discriminative Model Calibration

```

Even deep learning models with high precision and recall can be poorly calibrated, and the overconfidence of these models has been noted [@arxiv:1706.04599; @arxiv:1807.00263].
We attempted to calibrate the best performing discriminative model so that we could directly use the emitted probabilities.
We examined the calibration of our existing model (Figure {@fig:discriminative_model_calibration}, blue line).
We found that the DaG and CtG edge types were, though not perfectly calibrated, were somewhat aligned with the ideal calibration lines.
The CbG and GiG edges were poorly calibrated and increasing model certainty did not always lead to an increase in precision.
Applying the calibration algorithm (orange line) did not appear to bring predictions in line with the ideal calibration line, but did capture some of the uncertainty in the GiG edge type.
For this reason we use the measured precision instead of the predicted probabilities when determining how many edges could be added to existing knowledge bases with specified levels of confidence.

![
Calibration plots for the discriminative model.

> A perfectly calibrated model would show a diagonal line (dashed).

Reviewer: ```suggestion
A perfectly calibrated model would follow the dashed diagonal line.
```

place the grid aurocs here for generative model

## Discriminator Model Builds Off Generative Model

> We evaluate the discriminator model's ability to build off of the generative model.

Reviewer 2: More description of what this means needed. Whenever discussing machine learning classifiers, let the user know what the observations are, what the features are, and what the status (outcome variable) is.

For each generative model, trained on sampled label functions, we trained a discriminative model.
We report the results of the discriminative model using area under the precision recall curve (AUPRC) (Figure {@fig:discriminative_model_auprc}) and area under the receiver operating curve (AUROC) (Figure {@fig:discriminative_model_auroc}).    

> A general trend is that performance increases when the discriminative model is trained using generative models design each edge type. 

Reviewer 2: grammer problem in later half of sentence

For example, the top-left most square depicts the discriminator model predicting Disease associates Gene (DaG) sentences, while randomly sampling label functions designed to predict the DaG relationship. 
The square towards the right depicts the discriminative model predicting DaG sentences, while randomly sampling label functions designed to predict the Compound treats Disease (CtD) relationship.
This pattern continues filling out the rest of the grid.
The last most column consists of pooling every relationship specific label function and proceeding as above.
](images/figures/label_sampling/disc_performance_test_set_auprc.png){#fig:discriminative_model_auprc}

## Discriminative Model Calibration
Deep learning models can be overconfident in their predictions. 
Based on this notion we calibrated the best performing discriminative model using a temperature scalling algorithm.
For Disease associates Gene (DaG) and Compound treats Disease (CtD) the model is calibrated for high prob predictions.
For Compound binds Gene (CbG) and Gene interacts Gene (GiG) the model does not calibrate well.

> The CbG model has the predicted probabilities shrinking into a domain of [0.2, 0.6].

Reviewer 2: Would "0.2–0.6" be more natural of english?

> The GiG model predictions were overestimatedm so the calibrated range shrink to be between 0.4 and 0.6.

Reviewer 2: overestimatedm

![
Calibration plots for the discriminative model.

> A perfectly calibrated model would show a diagnoal line (dashed).

Reviewer 2: diagnoal
